### ISSUE TITLE
Mapping committers to authors

### ISSUE THREAD

**Issue by mfenner** [force11/force11-sciwg#8]
How do we generate an author list from a list of people who committed to software source code? There is a technical aspect, e.g. fetching real names from GitHub usernames. But there is also the bigger question of who should be considered an author in the software citation sense.

**Comment by brucellino** [comment]
Perhaps there could be some standard convention, whereby authors are kept in `authors.yml` and contributors are pulled from the git logs ? 
Don't mean to barge in on a discussion, but I have been having to deal with this recently.

**Comment by mfenner** [comment]
Thanks a lot Bruce. We are just starting the discussion on how to do this. Related to using `authors.yml` is the idea of a more generic CITATION or `codemeta.json` file that also holds other metadata. See https://github.com/force11/force11-sciwg/issues/2 and https://github.com/force11/force11-sciwg/issues/4.

**Comment by mfenner** [comment]
In the GitHub -> Zenodo workflow adding author names is a manual step via a web form, which gives the flexibility to not simply use the committers, but makes automation a bit more difficult.

**Comment by brucellino** [comment]
Yes, I have been exposed to this issue exclusively through the Github->Zenodo workflow. It automatically fills the form with the contributors list, which I need to go and fix every time. It also includes bots and agents which have "contributed" pull requests from CI/CD platforms. I guess maintainers would like to deal with that in the repository itself (at least, I would), so that you can prepare the release before requesting a DOI and deposition to a repository, pushing all of the metadata that is required at submit time. This is not just a means for saving time through automation, but also promotes consistency and proper attribution.

**Comment by mfenner** [comment]
I totally agree. And I think a file containing author information might be the easiest way to accomplish this. Maybe @lnielsen can provide feedback of what Zenodo is planning to do, e.g. whether it could fetch such a file in the GitHub -> Zenodo workflow.

**Comment by danielskatz** [comment]
Regarding authors, while it's good to use automation where possible, we should also try to be flexible and consider different situations, such as:
1 - the software that we are citing is not open source
2 - there are "authors" who did not make any commits
so perhaps automation could be used to create a set of "default" authors, and then the submitter could edit/change this before submission.  Or if it's a file in a repo, the automated tool could fill in the initial version of the file, which someone else could alter.

A problem that underlies this is that we don't really have a definition of "author" that is consistent in all cases.

Perhaps we can use "contributor", where a contributor could be a committer of software, an author of documentation, or someone else who has made a contribution to the software in a different form.

**Comment by lnielsen** [comment]
@mfenner Yes, we can already fetch a file with metadata. If you don't do anything, the GitHub-Zenodo integration will do a **best guess** based on the GitHub committer statistics. You can however also currently stick a ``.zenodo.json`` in the repository, which we'll read and grab the metadata from (available since Sep 2016).

![screen shot 2017-05-15 at 08 20 52](https://cloud.githubusercontent.com/assets/1698163/26044906/87b0307e-3947-11e7-819f-be43763c53b9.png)

I think any automatic algorithm for determining authorship based on committer data is doomed to 
fail. Zenodo's best guess I think actually only works in the cases where authors don't care (i.e. it doesn't work)! The only real  way IMHO is to specify the authors explicitly in the repository (i.e. something like #2 or #4 - also a generic solution is definitely preferably over the current ``.zenodo.json``).

That said, I think this issue can still be useful to discuss, as we can use it to create tools that help authors generate a ``CITATION`` or ``codemeta.json`` file for their repository. We should however just be clear on that #2 and #4 is the real solution for declaring authorship, and this issue is just a means to achieving #2 or #4.

I think the rest boils down to two key problems:

1) Who: Identifying who should be in the author list
2) Order: Identifying the order of the authors in the list.

**Who**
Identifying committers in the commit history is the first issue. As @brucellino mentions you can have bots in the history, however you can also have humans in the commit message declared as e.g. ``Co-authored-by: Person <email>`` that makes this difficult.

**Order**
Quantity vs quality: It might be that the core of a piece of software is an algorithm of 100 lines (i.e. researcher inventing the algorithm perhaps didn't even write it), and the rest of the repository 10k+ lines is readme, license headers, tests, .... I.e. order of authors can be pretty difficult to try to get.

Once above two issues is solved, then as @mfenner mentions then you also have problems translating a committer into an author name.

**Comment by mfenner** [comment]
Very cool. I wasn't aware of the `.zenodo.json` file. It looks like DataCite metadata in json format, plus additional properties for Zenodo (e.g. `communities`). 

BTW, since earlier this month DataCite content negotiation supports DataCite metadata in json format. This is still evolving, but you get something similar to the JSON in `.zenodo.json`:

```
curl -LH "Accept: application/vnd.datacite.datacite+json" https://doi.org/10.5281/ZENODO.46440 
```

Would Zenodo be interested in supporting `codemeta.json` as an input format? Which would mean that we should make sure it supports the metadata you need. But this discussion really belongs into #4.

It seems that there is agreement that the best solution for providing author names is a file in the code repository. There is another option in addition to CITATION #2 and codemeta.json #4: software repositories such as CRAN, PyPi , npmor Rubygems require a special metadata file and the author information in there is probably very close to what we want for data citation.

**Comment by lnielsen** [comment]
> supports DataCite metadata in json format. 

Cool. If the JSON format stabilises I would be interested in supporting it in the Python datacite package: https://datacite.readthedocs.io/en/latest/

> Would Zenodo be interested in supporting codemeta.json as an input format? 

Yes, actually we more than interested. It's part of our short-term roadmap to do this.

**Comment by mfenner** [comment]
> Yes, actually we more than interested. It's part of our short-term roadmap to do this.

Very nice. 

It is on our medium-term roadmap to allow the registration of DOIs via metadata that are not DataCite XML, and that includes DataCite JSON, but also codemeta and schema.org.

**Comment by brucellino** [comment]
@lnielsen the `.zenodo.json` is a beautiful thing. I spent a lot of time "fixing" whatever github sent to Zenodo, and including things which Github doesn't know about (Zenodo communities, funding, etc). This makes thing much much easier and potentially more accurate. 

Is there a best-practice document in how to create that JSON, and a full schema (forgive me if I've missed it in the previous discussion). 

I then have a bigger question about standardisation - this seems like a zenodo-specific helper. What can other kinds of repositories do to consume a more generic citation file in the repository ? (perhaps requires a new issue to discuss ?)

**Comment by lnielsen** [comment]
> Is there a best-practice document in how to create that JSON, and a full schema (forgive me if I've missed it in the previous discussion).

If you already made a release, your repository will have a JSON you can start from (as in the screenshot above). The JSON is the same as our [REST API](http://developers.zenodo.org/#representation)

> I then have a bigger question about standardisation - this seems like a zenodo-specific helper. What can other kinds of repositories do to consume a more generic citation file in the repository ? (perhaps requires a new issue to discuss ?)

Wait for the result of #2 and #4 :-) On Zenodo we really don't want a Zenodo-specific JSON as we have now. We'll much rather have a generic one based on either #2 or #4. We were holding off on the ``.zenodo.json`` for quite some time, but in the end decided to publish due to all the problems with correct authorship. We'll be hiding support for the ``.zenodo.json`` as soon as we have better generic solution.
